File: app/src/main/java/io/treehouses/remote/adapter/ServiceCardAdapter.java
package io.treehouses.remote.adapter;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

import io.treehouses.remote.Fragments.ServiceCardFragment;
import io.treehouses.remote.pojo.ServiceInfo;

public class ServiceCardAdapter extends FragmentStatePagerAdapter {
    private ArrayList<ServiceInfo> data;

    public ServiceCardAdapter(FragmentManager fm, ArrayList<ServiceInfo> data) {
        super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        this.data = removeHeaders(data);

    }

    private ArrayList<ServiceInfo> removeHeaders(ArrayList<ServiceInfo> data) {
        ArrayList<ServiceInfo> tmp = new ArrayList<>(data);
        Iterator<ServiceInfo> iterator = tmp.iterator();
        while (iterator.hasNext()) {
            if (iterator.next().isHeader()) iterator.remove();
        }
        return tmp;
    }

    @Override
    public int getCount() {
        return data.size();
    }
    public ArrayList<ServiceInfo> getData() {
        return data;
    }


    @NonNull
    @Override
    public Fragment getItem(int position) {
        Bundle b = new Bundle();
        b.putSerializable("serviceData",data.get(position));
        Fragment f = new ServiceCardFragment();
        f.setArguments(b);
        return f;
    }

    @Override
    public int getItemPosition(Object o) {
        return POSITION_NONE;
    }


    @Override
    public void notifyDataSetChanged() {
        Collections.sort(data);
        super.notifyDataSetChanged();
    }


}

File: app/src/main/kotlin/io/treehouses/remote/adapter/ServiceCardAdapter.kt
package io.treehouses.remote.adapter

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.PagerAdapter
import io.treehouses.remote.Fragments.ServiceCardFragment
import io.treehouses.remote.pojo.ServiceInfo
import java.util.*

class ServiceCardAdapter(fm: FragmentManager?, data: MutableList<ServiceInfo>) : FragmentStatePagerAdapter(fm!!, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
    var data: ArrayList<ServiceInfo>
    private fun removeHeaders(data: MutableList<ServiceInfo>): ArrayList<ServiceInfo> {
        val tmp = ArrayList(data)
        val iterator = tmp.iterator()
        while (iterator.hasNext()) {
            if (iterator.next().isHeader) iterator.remove()
        }
        return tmp
    }

    override fun getCount(): Int {
        return data.size
    }

    override fun getItem(position: Int): Fragment {
        val b = Bundle()
        b.putSerializable("serviceData", data[position])
        val f: Fragment = ServiceCardFragment()
        f.arguments = b
        return f
    }

    override fun getItemPosition(o: Any): Int {
        return PagerAdapter.POSITION_NONE
    }

    override fun notifyDataSetChanged() {
        this.data = removeHeaders(data)
        data.sort()
        super.notifyDataSetChanged()
    }

    init {
        this.data = removeHeaders(data)
    }
}